lib/handler/retry-handler.js

```diff
@@ -59,292 +70,227 @@ class RetryHandler {
-    if (Array.isArray(methods) && !methods.includes(method)) {
-      cb(err)
-      return
-    }
-
-    // If a set of status code are provided and the current status code is not in the list
-    if (
-      statusCode != null &&
-      Array.isArray(statusCodes) &&
-      !statusCodes.includes(statusCode)
-    ) {
-      cb(err)
-      return
-    }
-
-    // If we reached the max number of retries
-    if (counter > maxRetries) {
-      cb(err)
-      return
-    }
-
-    let retryAfterHeader = headers?.['retry-after']
-    if (retryAfterHeader) {
-      retryAfterHeader = Number(retryAfterHeader)
-      retryAfterHeader = Number.isNaN(retryAfterHeader)
-        ? calculateRetryAfterHeader(retryAfterHeader)
-        : retryAfterHeader * 1e3 // Retry-After is in seconds
-    }
-
-    const retryTimeout =
-      retryAfterHeader > 0
-        ? Math.min(retryAfterHeader, maxTimeout)
-        : Math.min(minTimeout * timeoutFactor ** (counter - 1), maxTimeout)
-
-    setTimeout(() => cb(null), retryTimeout)
-  }
-
   onHeaders (statusCode, rawHeaders, resume, statusMessage) {
     const headers = parseHeaders(rawHeaders)
 
-    this.retryCount += 1
-
     if (statusCode >= 300) {
-      if (this.retryOpts.statusCodes.includes(statusCode) === false) {
-        return this.handler.onHeaders(
-          statusCode,
-          rawHeaders,
-          resume,
-          statusMessage
-        )
-      } else {
-        this.abort(
-          new RequestRetryError('Request failed', statusCode, {
-            headers,
-            count: this.retryCount
-          })
-        )
-        return false
-      }
-    }
-
-    // Checkpoint for resume from where we left it
-    if (this.resume != null) {
-      this.resume = null
-
-      if (statusCode !== 206) {
-        return true
+      if (this.#retryOpts.statusCodes.includes(statusCode)) {
+        // TODO (fix): Remove this path and delegate to different interceptor which will read the error body?
+        const message = `Response status code ${statusCode}${statusMessage ? `: ${statusMessage}` : ''}`
+        throw new ResponseStatusCodeError(message, statusCode, headers)
       }
 
-      const contentRange = parseRangeHeader(headers['content-range'])
-      // If no content range
-      if (!contentRange) {
-        this.abort(
-          new RequestRetryError('Content-Range mismatch', statusCode, {
-            headers,
-            count: this.retryCount
-          })
-        )
-        return false
+      this.#headersSent = true
+      return this.#handler.onHeaders(statusCode, rawHeaders, resume, statusMessage)
+    } else if (this.#reason == null) {
+      assert(this.#etag = null)
+      assert(this.#pos == null)
+      assert(this.#end == null)
+      assert(this.#headersSent === false)
+
+      if (headers.trailer) {
+        this.#headersSent = true
+        return this.#handler.onHeaders(statusCode, rawHeaders, resume, statusMessage)
```

Comment:
Never thought about the `trailers`, basically if trailers are hinted we just forward do not retry?

Shall we add some flag for implementers to maybe bypass it?
I see the benefit on this